aiida/orm/nodes/data/singlefile.py

```diff
@@ -50,15 +51,35 @@ def __init__(self, file, filename: str | pathlib.Path | None = None, **kwargs):
             self.set_file(file, filename=filename)
 
     @property
-    def filename(self):
+    def filename(self) -> str:
         """Return the name of the file stored.
 
         :return: the filename under which the file is stored in the repository
         """
         return self.base.attributes.get('filename')
 
+    @t.overload
     @contextlib.contextmanager
-    def open(self, path=None, mode='r'):
+    def open(self, path: str, mode: t.Literal['r']) -> t.Iterator[t.TextIO]:
+        ...
+
+    @t.overload
+    @contextlib.contextmanager
+    def open(self, path: None, mode: t.Literal['r']) -> t.Iterator[t.TextIO]:
+        ...
+
+    @t.overload
+    @contextlib.contextmanager
+    def open(self, path: str, mode: t.Literal['rb']) -> t.Iterator[t.BinaryIO]:
+        ...
+
+    @t.overload
+    @contextlib.contextmanager
+    def open(self, path: None, mode: t.Literal['rb']) -> t.Iterator[t.BinaryIO]:
+        ...
+
+    @contextlib.contextmanager  # type: ignore[misc]
+    def open(self, path: str | None = None, mode: t.Literal['r', 'rb'] = 'r') -> t.Iterator[t.BinaryIO | t.TextIO]:
```

Comment:
From the example here https://docs.python.org/3/library/typing.html#typing.overload, seems it is not needed to provide typing for the actual implementation? But I guess providing it did nothing harm.